lib/src/ast.dart

```diff
@@ -1,5 +1,16 @@
 /// Base interface for queries.
 abstract class Query {
+  const Query({
+    required this.startIndex,
```

Comment:
Having thought of it, I think I'd rather introduce a `SourcePosition` or `SourceSpan` class with `start` and `end` fields, and only have either `position`, `sourcePosition`, `span` or `sourceSpan` field here.
Kind of like in https://pub.dev/documentation/source_span/latest/source_span/SourceSpan-class.html, but less complex. wdyt?